Creation of EVC scheduler can eventually leads to race condition

As part of AmLight end-to-end tests daily execution, the creation of EVC scheduler can eventually leads to race condition and lost of user request, as can be shown below:

```
_____________________ TestE2EMefEline.test_delete_schedule _____________________

self = <tests.test_e2e_12_mef_eline.TestE2EMefEline instance at 0x7f0c79252170>
circuit_id = '6d77ec4583034d'

   def test_delete_schedule(self, circuit_id):
       """ Test to delete a scheduler. """

       # Schedule by frequency every minute
       payload = {
           "circuit_id": circuit_id,
           "schedule": {
               "frequency": "* * * * *"
           }
       }

       # Create circuit schedule
       api_url = KYTOS_API + '/mef_eline/v2/evc/schedule'
       response = requests.post(api_url, json=payload)
       assert response.status_code == 201

       # Verify the list of schedules
       api_url = KYTOS_API + '/mef_eline/v2/evc/schedule/'
       response = requests.get(api_url)
       assert response.status_code == 200

     data = response.json()[0]
E       IndexError: list index out of range

tests/test_e2e_12_mef_eline.py:188: IndexError
======== 1 failed, 164 passed, 20 xfailed, 4 xpassed in 8022.95 seconds ========

```

Debug logs for this execution shows that the EVC was created/deployed right before the user request a schedule, which can be a indicator that user data was overwritten by concurrent execution:

```
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:148:  create_circuit /v2/evc/
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:INFO evc:131:  EVC 6d77ec4583034d was synced to the storehouse.
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU evc:454:  Path is empty.
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:INFO evc:131:  EVC 6d77ec4583034d was synced to the storehouse.
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:INFO evc:519:  EVC(6d77ec4583034d, my evc1) was deployed.
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:217:  create_circuit result {'circuit_id': '6d77ec4583034d'} 201
Dec 16 05:48:39 xxxxx werkzeug:INFO _internal:113:  127.0.0.1 - - [16/Dec/2021 05:48:39] "#033[37mPOST /api/kytos/mef_eline/v2/evc/ HTTP/1.1#033[0m" 201 -
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:108:  get_circuit /v2/evc/6d77ec4583034d
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:119:  get_circuit result {'id': '6d77ec4583034d', 'name': 'my evc1', 'uni_a': {'interface_id': '00:00:00:00:00:00:00:01:1', 'tag': None}, 'uni_z': {'interface_id': '00:00:00:00:00:00:00:01:2', 'tag': None}, 'start_date': '2021-12-16T05:48:39', 'end_date': None, 'queue_id': None, 'bandwidth': 0, 'primary_links': [], 'backup_links': [], 'current_path': [], 'primary_path': [], 'backup_path': [], 'dynamic_backup_path': False, 'metadata': {}, 'request_time': '2021-12-16T05:48:39', 'creation_time': '2021-12-16T05:48:39', 'owner': None, 'circuit_scheduler': [], 'active': True, 'enabled': True, 'archived': False, 'priority': -1} 200
Dec 16 05:48:39 xxxxx werkzeug:INFO _internal:113:  127.0.0.1 - - [16/Dec/2021 05:48:39] "#033[37mGET /api/kytos/mef_eline/v2/evc/6d77ec4583034d HTTP/1.1#033[0m" 200 -
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:431:  create_schedule /v2/evc/schedule/
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:INFO evc:131:  EVC 6d77ec4583034d was synced to the storehouse.
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:488:  create_schedule result {'id': '537c1323e9d346989f1a7c4b3b4d70d9', 'action': 'create', 'frequency': '* * * * *'} 201
Dec 16 05:48:39 xxxxx werkzeug:INFO _internal:113:  127.0.0.1 - - [16/Dec/2021 05:48:39] "#033[37mPOST /api/kytos/mef_eline/v2/evc/schedule HTTP/1.1#033[0m" 201 -
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:393:  list_schedules /v2/evc/schedule
Dec 16 05:48:39 xxxxx kytos.napps.kytos/mef_eline:DEBU main:411:  list_schedules result [] 200
Dec 16 05:48:39 xxxxx werkzeug:INFO _internal:113:  127.0.0.1 - - [16/Dec/2021 05:48:39] "#033[37mGET /api/kytos/mef_eline/v2/evc/schedule/ HTTP/1.1#033[0m" 200 -
```

As we can see the EVC is persisted to the storehouse (from the EVC creation routine) right after the scheduler displays its results, which is a strong indicator that a race condition leaded to the test failure.

--- a/storehouse.py
+++ b/storehouse.py
@@ -77,6 +77,8 @@ class StoreHouse:
                    'data': {}}
         name = 'kytos.storehouse.retrieve'
         event = KytosEvent(name=name, content=content)
+        self._lock.acquire()  # Lock to avoid race condition
+        log.debug(f'Lock {self._lock} acquired.')
         self.controller.buffers.app.put(event)
         log.debug(f'Retrieve box with {box_id} from {self.namespace}.')
 
@@ -86,6 +88,8 @@ class StoreHouse:
             log.error(f'Box {data.box_id} not found in {self.namespace}.')
 
         self.box = data
+        self._lock.release()
+        log.debug(f'Lock {self._lock} released.')
         log.debug(f'Box {self.box.box_id} was loaded from storehouse.')
 
     def save_evc(self, evc):

--- a/tests/unit/test_main.py
+++ b/tests/unit/test_main.py
@@ -918,6 +918,90 @@ class TestMain(TestCase):
                          response_json["action"])
         self.assertIsNotNone(response_json["id"])
 
+        # Case 2: there is no schedule
+        payload = {
+              "circuit_id": "cc:cc:cc",
+              "schedule": {
+                "frequency": "1 * * * *",
+                "action": "create"
+              }
+            }
+        response = api.post(url, data=json.dumps(payload),
+                            content_type='application/json')
+        self.assertEqual(response.status_code, 201)
+
+    def test_create_schedule_invalid_request(self):
+        """Test create schedule API with invalid request."""
+        evc1 = MagicMock()
+        self.napp.circuits = {'bb:bb:bb': evc1}
+        api = self.get_app_test_client(self.napp)
+        url = f'{self.server_name_url}/v2/evc/schedule/'
+
+        # case 1: empty post
+        response = api.post(url, data="")
+        self.assertEqual(response.status_code, 415)
+
+        # case 2: content-type not specified
+        payload = {
+            "circuit_id": "bb:bb:bb",
+            "schedule": {
+                "frequency": "1 * * * *",
+                "action": "create"
+            }
+        }
+        response = api.post(url, data=json.dumps(payload))
+        self.assertEqual(response.status_code, 415)
+
+        # case 3: not a dictionary
+        payload = []
+        response = api.post(url, data=json.dumps(payload),
+                            content_type='application/json')
+        self.assertEqual(response.status_code, 400)
+
+        # case 4: missing circuit id
+        payload = {
+            "schedule": {
+                "frequency": "1 * * * *",
+                "action": "create"
+            }
+        }
+        response = api.post(url, data=json.dumps(payload),
+                            content_type='application/json')
+        self.assertEqual(response.status_code, 400)
+
+        # case 5: missing schedule
+        payload = {
+            "circuit_id": "bb:bb:bb"
+        }
+        response = api.post(url, data=json.dumps(payload),
+                            content_type='application/json')
+        self.assertEqual(response.status_code, 400)
+
+        # case 6: invalid circuit
+        payload = {
+            "circuit_id": "xx:xx:xx",
+            "schedule": {
+                "frequency": "1 * * * *",
+                "action": "create"
+            }
+        }
+        response = api.post(url, data=json.dumps(payload),
+                            content_type='application/json')
+        self.assertEqual(response.status_code, 404)
+
+        # case 7: archived or deleted evc
+        evc1.archived.return_value = True
+        payload = {
+            "circuit_id": "bb:bb:bb",
+            "schedule": {
+                "frequency": "1 * * * *",
+                "action": "create"
+            }
+        }
+        response = api.post(url, data=json.dumps(payload),
+                            content_type='application/json')
+        self.assertEqual(response.status_code, 403)
+
     @patch('apscheduler.schedulers.background.BackgroundScheduler.remove_job')
     @patch('napps.kytos.mef_eline.storehouse.StoreHouse.get_data')
     @patch('napps.kytos.mef_eline.scheduler.Scheduler.add')
@@ -1125,6 +1209,15 @@ class TestMain(TestCase):
         response = api.delete(url)
 
         self.assertEqual(response.status_code, 403, response.data)
+
+    @patch('napps.kytos.mef_eline.main.Main._find_evc_by_schedule_id')
+    def test_delete_schedule_not_found(self, mock_find_evc_by_sched):
+        """Test delete a circuit schedule - unexisting."""
+        mock_find_evc_by_sched.return_value = (None, False)
+        api = self.get_app_test_client(self.napp)
+        url = f'{self.server_name_url}/v2/evc/schedule/1'
+        response = api.delete(url)
+        self.assertEqual(response.status_code, 404)
 
     @patch('requests.post')
     @patch('napps.kytos.mef_eline.scheduler.Scheduler.add')

--- a/tests/unit/test_storehouse.py
+++ b/tests/unit/test_storehouse.py
@@ -56,3 +56,53 @@ class TestStoreHouse(TestCase):
         self.storehouse.box.box_id = 2
         self.storehouse.get_data()
         get_stored_box_mock.assert_called_once_with(2)
+
+    @patch('napps.kytos.mef_eline.storehouse.log')
+    def test_create_box_callback(self, log_mock):
+        # pylint: disable=protected-access
+        """Test _create_box_callback method."""
+        data = Mock()
+        data.box_id = 1
+        self.storehouse._create_box_callback('event', data, None)
+        log_mock.error.assert_not_called()
+        log_mock.debug.assert_called_once()
+
+        # test with error
+        log_mock.debug.call_count = 0
+        self.storehouse._create_box_callback('event', data, 'error')
+        log_mock.error.assert_called_once()
+        log_mock.debug.assert_called_once()
+
+    @patch('napps.kytos.mef_eline.storehouse.StoreHouse.get_stored_box')
+    @patch('napps.kytos.mef_eline.storehouse.StoreHouse.create_box')
+    def test_get_or_create_box(self, create_box_mock, get_stored_box_mock):
+        """Test _get_or_create_a_box_from_list_of_boxes method."""
+        # pylint: disable=protected-access
+        self.storehouse._get_or_create_a_box_from_list_of_boxes(
+            'event', [2], None
+        )
+        get_stored_box_mock.assert_called_once_with(2)
+
+        self.storehouse._get_or_create_a_box_from_list_of_boxes(
+            'event', None, 'error'
+        )
+        create_box_mock.assert_called_once()
+
+    @patch('napps.kytos.mef_eline.storehouse.log')
+    def test_get_box_callback(self, log_mock):
+        # pylint: disable=protected-access
+        """Test _get_box_callback method."""
+        self.storehouse._lock = Mock()
+        data = Mock()
+        data.box_id = 1
+        self.storehouse._get_box_callback('event', data, None)
+        self.storehouse._lock.release.assert_called_once()
+        log_mock.error.assert_not_called()
+
+        # test with error
+        self.storehouse._lock.release.call_count = 0
+        log_mock.debug.call_count = 0
+        self.storehouse._get_box_callback('event', data, 'error')
+        self.storehouse._lock.release.assert_called_once()
+        log_mock.error.assert_called_once()
+        self.assertEqual(log_mock.debug.call_count, 2)

